kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/EnvelopeEncryption.java

```diff
@@ -59,39 +64,121 @@ record KmsCacheConfig(
+            Kms<K, E> kms = kmsPlugin.buildKms(configuration.kmsConfig());
+            kms = InstrumentedKms.wrap(kms, kmsMetrics);
+            ExponentialJitterBackoffStrategy backoffStrategy = new ExponentialJitterBackoffStrategy(Duration.ofMillis(500), Duration.ofSeconds(5), 2d,
+                    ThreadLocalRandom.current());
+            kms = ResilientKms.wrap(kms, context.eventLoop(), backoffStrategy, 3);
+            return wrapWithCachingKms(configuration, kms);
+        }
+
+        @NonNull
+        private static <K, E> Kms<K, E> wrapWithCachingKms(Config configuration, Kms<K, E> resilientKms) {
+            KmsCacheConfig config = configuration.kmsCache();
+            return CachingKms.wrap(resilientKms, config.decryptedDekCacheSize, config.decryptedDekExpireAfterAccessDuration, config.resolvedAliasCacheSize,
+                    config.resolvedAliasExpireAfterWriteDuration, config.resolvedAliasRefreshAfterWriteDuration);
+        }
+
+        PerEventLoopState<K, E> stateFor(FilterFactoryContext context) {
+            ScheduledExecutorService eventLoop = context.eventLoop();
+            if (eventLoop == null) {
+                throw new IllegalStateException("eventloop on context is null");
+            }
+            return stateMap.computeIfAbsent(eventLoop, exec -> new PerEventLoopState<>(context, this));
+        }
+
+        public Kms<K, E> kms() {
+            return kms;
+        }
+
+        public Map<ScheduledExecutorService, PerEventLoopState<K, E>> stateMap() {
+            return stateMap;
+        }
+
+        public Config config() {
+            return config;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+            if (obj == null || obj.getClass() != this.getClass()) {
+                return false;
+            }
+            var that = (ApplicationWideState) obj;
+            return Objects.equals(this.kms, that.kms) &&
+                    Objects.equals(this.stateMap, that.stateMap) &&
+                    Objects.equals(this.config, that.config);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(kms, stateMap, config);
+        }
+
+        @Override
+        public String toString() {
+            return "ApplicationWideState[" +
+                    "kms=" + kms + ", " +
+                    "stateMap=" + stateMap + ", " +
+                    "config=" + config + ']';
+        }
+
+        public DekAllocator<K, E> getDekAllocator() {
+            return dekAllocator;
+        }
     }
 
-    @NonNull
-    @Override
-    public EnvelopeEncryptionFilter<K> createFilter(FilterFactoryContext context, Config configuration) {
-        Kms<K, E> kms = buildKms(context, configuration);
+    public record PerEventLoopState<K, E>(KeyManager<K> keyManager, TopicNameBasedKekSelector<K> selector) {
```

Comment:
Yeap, also depending on the identity of an Executor is unsafe. I'm not sure how Kroxy could cleanly support this pattern since it's dependent on the concept of Filter instances having a 1:1 relationship with a thread. You'd have to  guarantee that it's to be associated with a single thread, then maybe we could have some method on `FilterFactory` like `Y initialiseFilterThreadState(...)` guaranteed to be called once per "filter-thread".

Maybe we're better off with a different approach like using Threadlocals from within the Filter to obtain a `keyManager` and `kekSelector`.

Reply:
Now I'm squinting at FilterHandler worried that the Filter work could be driven by whatever uncontrolled thread is completing a previous filter's future.

Reply:
think we should be switching back to the channel executor ASAP

Reply:
Have had a look and it's not a problem between Filter instances, they each have a FilterHandler that is driven by the eventloop on write. But it's potentially a problem if we have received a batch of requests (like zero-ack Produce Requests), and they are enqueued for a single filter, they could get dispatched on an unexpected thread.